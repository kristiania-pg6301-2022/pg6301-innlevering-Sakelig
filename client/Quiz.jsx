import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

const Quiz = ({ quizApi }) => {
  const [question, setQuestion] = useState("");
  let [answer, setAnswer] = useState("");
  let id = question.id;

  const navigate = useNavigate();

  useEffect(async () => {
    setQuestion(await quizApi.getQuestion());
  }, []);

  async function handleSubmit(e) {
    e.preventDefault();

    const data = await quizApi.postAnswer({ id, answer });

    if (data.result === true) {
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }

  let test2 = "";

  if (question) {
    test2 = Object.keys(question.answers)
      .filter((x) => question.answers[x])
      .map((x) => (
        <div key={x}>
          <form onSubmit={handleSubmit}>
            <button
              name={"answer"}
              value={question.answers[x]}
              onClick={() => setAnswer(x)}
            >
              {question.answers[x]}
            </button>
          </form>
        </div>
      ));
  } else {
    test2 = "question is unidentified :c";
  }

  return (
    <div>
      <h1>Quiz</h1>
      <div>
        <h2>{question.question}</h2>
        <div>{test2}</div>
        {/*answer && isCorrectAnswer(question, answer) ? navigate("/answer/correct") : navigate("/answer/wrong")*/}
      </div>
    </div>
  );
};

export default Quiz;
