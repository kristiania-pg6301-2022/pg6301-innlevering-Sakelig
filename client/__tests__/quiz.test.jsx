import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";
import Quiz from "../Quiz";
import ReactDOM from "react-dom";
import React from "react";

describe("Quizpage", () => {
  it("show loaded quizpage", async () => {
    const getQuestion = () => ({
      id: 974,
      question:
        "What is the correct JavaScript syntax to change the content of the HTML element below?",
      answers: {
        answer_a: '#demo.innerHTML = "Hello World!";',
        answer_b: 'document.getElementById("demo").innerHTML = "Hello World!";',
        answer_c: 'document.getElement("p").innerHTML = "Hello World!";',
        answer_d: 'document.getElementByName("p").innerHTML = "Hello World!";',
        answer_e: null,
        answer_f: null,
      },
    });

    const element = document.createElement("div");

    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Quiz quizApi={{ getQuestion }} />
        </MemoryRouter>,
        element
      );
    });
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("can simulate correct answer", async () => {
    const getQuestion = () => ({
      id: 974,
      question:
        "What is the correct JavaScript syntax to change the content of the HTML element below?",
      answers: {
        answer_a: '#demo.innerHTML = "Hello World!";',
        answer_b: 'document.getElementById("demo").innerHTML = "Hello World!";',
        answer_c: 'document.getElement("p").innerHTML = "Hello World!";',
        answer_d: 'document.getElementByName("p").innerHTML = "Hello World!";',
        answer_e: null,
        answer_f: null,
      },
    });

    let postAnswer = jest.fn().mockReturnValue({ result: true });

    const element = document.createElement("div");

    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Quiz quizApi={{ getQuestion, postAnswer }} />
        </MemoryRouter>,
        element
      );
    });

    Simulate.submit(element.querySelector("form"));

    /*
    - Cant find answer because button that sets it is never called
    - this going through causes error since rest of code is trying to
      use the result that is normally recived from the function

    - Not sure why act is acting up :frog:
     */

    expect(postAnswer).toBeCalledWith({
      answer: "",
      id: 974,
    });
  });

  it("can simulate wrong answer", async () => {
    const getQuestion = () => ({
      id: 974,
      question:
        "What is the correct JavaScript syntax to change the content of the HTML element below?",
      answers: {
        answer_a: '#demo.innerHTML = "Hello World!";',
        answer_b: 'document.getElementById("demo").innerHTML = "Hello World!";',
        answer_c: 'document.getElement("p").innerHTML = "Hello World!";',
        answer_d: 'document.getElementByName("p").innerHTML = "Hello World!";',
        answer_e: null,
        answer_f: null,
      },
    });

    let postAnswer = jest.fn().mockReturnValue({ result: false });

    const element = document.createElement("div");

    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Quiz quizApi={{ getQuestion, postAnswer }} />
        </MemoryRouter>,
        element
      );
    });

    Simulate.submit(element.querySelector("form"));

    /*
    - Cant find answer because button that sets it is never called
    - this going through causes error since rest of code is trying to
      use the result that is normally recived from the function

    - Not sure why act is acting up :frog:
     */

    expect(postAnswer).toBeCalledWith({
      answer: "",
      id: 974,
    });
  });
});
