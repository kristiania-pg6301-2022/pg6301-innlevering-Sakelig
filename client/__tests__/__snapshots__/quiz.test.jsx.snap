// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quizpage show loaded quizpage 1`] = `"<div><h1>Quiz</h1><div><h2>What is the correct JavaScript syntax to change the content of the HTML element below?</h2><div><div><form><button name=\\"answer\\" value=\\"#demo.innerHTML = &quot;Hello World!&quot;;\\">#demo.innerHTML = \\"Hello World!\\";</button></form></div><div><form><button name=\\"answer\\" value=\\"document.getElementById(&quot;demo&quot;).innerHTML = &quot;Hello World!&quot;;\\">document.getElementById(\\"demo\\").innerHTML = \\"Hello World!\\";</button></form></div><div><form><button name=\\"answer\\" value=\\"document.getElement(&quot;p&quot;).innerHTML = &quot;Hello World!&quot;;\\">document.getElement(\\"p\\").innerHTML = \\"Hello World!\\";</button></form></div><div><form><button name=\\"answer\\" value=\\"document.getElementByName(&quot;p&quot;).innerHTML = &quot;Hello World!&quot;;\\">document.getElementByName(\\"p\\").innerHTML = \\"Hello World!\\";</button></form></div></div></div></div>"`;
